### PR TITLE
PanZoom performance improvement

### LINKED ISSUE #2352: Scatter Plot pan/zoom is slow
Maybe it was improved by https://github.com/palantir/plottable/pull/2337 , but it needs double checking.

Spec:
` [...] we are having to limit to 5,000 dots for scatterplots so that they stay usable, we’d like to get upwards around 20,000 – 30,000 dots per plot eventually.
`

### PR DESCRIPTION
**Pull request by aicioara**
Closes #2352

**Description**: If we are having a ScatterPlot with a lot of data points, every time we update the scale (which happens on every single pan or zoom), all the `DOM` elements corresponding to the points get modified (in particular translated). This is a costly operation and can be improved by translating the parent `<g>` and letting the drawing engine do the heavy lifting. 

**Problem1**: We cannot account nicely for scaling (when zooming) here. That is because the scale on the top element also modifies the size of the points (and that is not acceptable). Therefore, we ran for a compromise: Scale the points (i.e. zoom badly and lose resolution) during the zoom and recompute all the points after no more interaction exists.

**Problem2**: Because of Problem1, we no longer have the same default behavior. Therefore, there is a switch `XYPlot.deferredRendering()`. 

**Unimplemented ideas**:
- <del>Maybe call performanceEnabled() something else, such that we can have finer grained performance enabling in the future </del>
- <del>Document everything, especially the new public method </del>
- <del>Extract the 500 ms in the setTimeout in some variable </del>
- Add tests?

---

#### Fiddles
Before: http://jsfiddle.net/rk0sx69o/
After: http://jsfiddle.net/rk0sx69o/7/

---

#### Numbers
Scatter Plot with 8000 points `6x increase`:
- Before: `175 ms` per update
- After: `3 ms` per update (capped by browser)

Old Scatter Plot with 8000 points has same performance as new Scatter Plot with 50 000

---

#### QE Notes: 
- Regression test: With the option `deferredRendering` turned off (that is the default). There should be absolutely no change.
- Feature test: With the option `deferredRendering` turned on, panning should behave the same. Zooming however, should behave weirdly (points / lines on plot appear to grow or shrink), but it should default to better quality after 0.5 seconds. The behavior should be the same as when zooming in/out on google map, i.e. low quality during zoom and get back to normal once we have data.
- Performance test: With the option `deferredRendering` turned on, panning and zooming should be 10x or more faster
- Additional notes: All plots except Pie Plot should be affected

### CODE REVIEW

File: plottable.d.ts [hunk 1]
```diff
@@ -2540,6 +2541,8 @@ declare module Plottable {
          * @param {Scale} yScale The y scale to use.
          */
         constructor();
+        performanceEnabled(): boolean;
+        performanceEnabled(performanceEnabled: boolean): XYPlot<X, Y>;
```
Comment (bluong) [inline plottable.d.ts]: What's the point of having `performanceEnabled()` be off?

**Comment by bluong** [comment]
One thing that I'm highly concerned with is that setting `performanceEnabled()` will give as a loss in features if it is set to true.  It makes sense to me that other features get slower at the cost of boosting the performance of panZoom but it doesn't make too much sense to me that features get axed due to the performance boost.

File: plottable.d.ts (hunk 1 above)
Comment (jtlan) [inline plottable.d.ts]: So the default behavior is the same as before.

**Comment by bluong** [comment]
In terms of user usage, let's say I find this performanceEnabled endpoint.  Aside from looking at the documentation (which is a step we generally want to somewhat avoid), there is nothing that says that the user shouldn't leave this at false.

File: plottable.d.ts (hunk 1 above)
Comment (aicioara) [inline plottable.d.ts]: Documented above

**Comment by jtlan** [comment]
@bluong wrote:
>One thing that I'm highly concerned with is that setting performanceEnabled() will give as a loss in features if it is set to true. It makes sense to me that other features get slower at the cost of boosting the performance of panZoom but it doesn't make too much sense to me that features get axed due to the performance boost.

The high-performance mode is there to offer a tradeoff: better speed at the cost of some precision / beauty during `Scale` updates. I think it makes sense to offer users the option to make that tradeoff.

@aicioara wrote:
>Maybe call performanceEnabled() something else, such that we can have finer grained performance enabling in the future

What about "`performanceMode()`"? Or something else where a `string` is passed in.

**Comment by bluong** [comment]
According to @aicioara however, it seems that the issue isn't some precision, it seems the plot will straight up not update nicely when zooming happens.

**Comment by aicioara** [comment]
@bluong 
Have you checked the fiddle?

File: src/plots/xyPlot.ts [hunk 2]
```diff
@@ -9,6 +9,19 @@ export class XYPlot<X, Y> extends Plot {
...
+  private _fastPanZoomOnYCallback: ScaleCallback<Scale<any, any>>;
+
+  private _fastPanZoomTimeoutReferenceX = 0;
+  private _fastPanZoomTimeoutReferenceY = 0;
+  private _fastPanZoomDeltaX = 0;
+  private _fastPanZoomDeltaY = 0;
+  private _fastPanZoomScaleX = 1;
+  private _fastPanZoomScaleY = 1;
+  private _fastPanZoomKnownDomainX: X[] = [null, null];
+  private _fastPanZoomKnownDomainY: Y[] = [null, null];
```
Comment (bluong) [inline src/plots/xyPlot.ts]: A bunch of the above variables allude to some `_fastPanZoom` idea but nowhere in the below code references panning or zooming.  All it seems to be is scale domain updating.

**Comment by aicioara** [comment]
The issue is the following: In order to keep the points to the same radius on a scatter plot (or the width of the lines on a line plot for what matters), we need to translate and not scale them. This means the translation is not the same for all points -> we cannot apply it to the parent DOM element -> we have to apply a different translate to every single point on the plot -> modify lots of DOM elements -> slow

File: src/plots/xyPlot.ts [hunk 3]
```diff
@@ -22,6 +35,95 @@ export class XYPlot<X, Y> extends Plot {
...
+        "translate(" + this._fastPanZoomDeltaX + ", " + this._fastPanZoomDeltaY + ")" +
+        "scale(" + this._fastPanZoomScaleX + ", " + this._fastPanZoomScaleY + ")");
+      clearTimeout(this._fastPanZoomTimeoutReferenceX);
+      this._fastPanZoomTimeoutReferenceX = setTimeout(() => {
+        this._fastPanZoomKnownDomainX = domain;
+        this._fastPanZoomDeltaX = 0;
+        this._fastPanZoomDeltaY = 0;
+        this.render();
+        this._renderArea.attr("transform", "translate(0, 0) scale(1, 1)");
+      }, 500);
```
Comment (bluong) [inline src/plots/xyPlot.ts]: Why is a timeout idea here?

File: src/plots/xyPlot.ts [hunk 4]
```diff
@@ -22,6 +35,95 @@ export class XYPlot<X, Y> extends Plot {
...
+
+    var domain = scale.domain();
+    this._fastPanZoomScaleX = (scale.scale(this._fastPanZoomKnownDomainX[1]) - scale.scale(this._fastPanZoomKnownDomainX[0])) /
+                              (scale.scale(domain[1]) - scale.scale(domain[0]));
+    this._fastPanZoomDeltaX = scale.scale(this._fastPanZoomKnownDomainX[0]) - scale.scale(domain[0]);
+
+    if (this._renderArea != null) {
+      this._renderArea.attr("transform",
+        "translate(" + this._fastPanZoomDeltaX + ", " + this._fastPanZoomDeltaY + ")" +
+        "scale(" + this._fastPanZoomScaleX + ", " + this._fastPanZoomScaleY + ")");
```
Comment (bluong) [inline src/plots/xyPlot.ts]: Where else are these instance variables use?  Why not wrap this function into a `_fastPanZoom` function()?

Also, this seems to be working with any type of scale updating.

File: src/plots/xyPlot.ts [hunk 5]
```diff
@@ -22,6 +35,95 @@ export class XYPlot<X, Y> extends Plot {
...
+      clearTimeout(this._fastPanZoomTimeoutReferenceY);
+      this._fastPanZoomTimeoutReferenceY = setTimeout(() => {
+        this._fastPanZoomKnownDomainY = domain;
+        this._fastPanZoomDeltaX = 0;
+        this._fastPanZoomDeltaY = 0;
+        this.render();
+        this._renderArea.attr("transform", "translate(0, 0) scale(1, 1)");
+      }, 500);
+    }
+  }
```
Comment (bluong) [inline src/plots/xyPlot.ts]: Lot's of duplicated logic it seems.  Can we factor out the logic.  Also, try to reduce the number of instance variables

File: src/plots/xyPlot.ts [hunk 6]
```diff
@@ -22,6 +35,95 @@ export class XYPlot<X, Y> extends Plot {
...
+
+  public performanceEnabled(): boolean;
+  public performanceEnabled(performanceEnabled: boolean): XYPlot<X, Y>;
```
Comment (bluong) [inline src/plots/xyPlot.ts]: If we want this endpoint, the documentation surrounding this will be ridiculously important (as well as the function name and the parameter name)

File: src/plots/xyPlot.ts (hunk 3 above)
Comment (aicioara) [inline src/plots/xyPlot.ts]: if no more PanZoom happens for 500 ms (will be checked with a designer), then do a render, basically restore the dimension of the points in the plot

File: src/plots/xyPlot.ts [hunk 7]
```diff
@@ -22,6 +35,95 @@ export class XYPlot<X, Y> extends Plot {
...
+      if (this.x() && this.x().scale) {
+        this._fastPanZoomKnownDomainX = this.x().scale.domain();
+        this.x().scale.onUpdate(this._fastPanZoomOnXCallback);
+        this.x().scale.offUpdate(this._renderCallback);
+      }
+      if (this.y() && this.y().scale) {
+        this._fastPanZoomKnownDomainY = this.y().scale.domain();
+        this.y().scale.onUpdate(this._fastPanZoomOnYCallback);
+        this.y().scale.offUpdate(this._renderCallback);
+      }
```
Comment (bluong) [inline src/plots/xyPlot.ts]: What happens if the scale is replaced after `performanceEnabled()` is set?

File: src/plots/xyPlot.ts (hunk 3 above)
Comment (jtlan) [inline src/plots/xyPlot.ts]: You get a similar effect in most map programs if you try to zoom in and out too fast; you'll initially see a lower-resolution version of what you zoomed in on, then a higher-res version will load.

File: src/plots/xyPlot.ts (hunk 6 above)
Comment (jtlan) [inline src/plots/xyPlot.ts]: Sounds like RFC material.

File: src/plots/xyPlot.ts [hunk 8]
```diff
@@ -22,6 +35,95 @@ export class XYPlot<X, Y> extends Plot {
 
     this._adjustYDomainOnChangeFromXCallback = (scale) => this._adjustYDomainOnChangeFromX();
     this._adjustXDomainOnChangeFromYCallback = (scale) => this._adjustXDomainOnChangeFromY();
+
+    this._fastPanZoomOnXCallback = (scale) => this._fastPanZoomOnX(scale);
+    this._fastPanZoomOnYCallback = (scale) => this._fastPanZoomOnY(scale);
```
Comment (bluong) [inline src/plots/xyPlot.ts]: Maybe `_fastRenderCallback`?

File: src/plots/xyPlot.ts (hunk 3 above)
Comment (bluong) [inline src/plots/xyPlot.ts]: We should maybe have this 500 be configurable for tests

**Comment by bluong** [comment]
Based on the demo you showed, you might want to do a survey to see if this should actually be the default / only behavior

File: src/plots/xyPlot.ts (hunk 6 above)
Comment (bluong) [inline src/plots/xyPlot.ts]: If you decide to keep this endpoint, maybe rename it to like `smoothScalingEnabled()`?

File: src/plots/xyPlot.ts (hunk 7 above)
Comment (aicioara) [inline src/plots/xyPlot.ts]: Good catch. It fails horribly for some reason

File: src/plots/xyPlot.ts (hunk 2 above)
Comment (aicioara) [inline src/plots/xyPlot.ts]: Fixed. Renamed to `lazyDomainChange[...]`

File: src/plots/xyPlot.ts (hunk 4 above)
Comment (aicioara) [inline src/plots/xyPlot.ts]: done

File: src/plots/xyPlot.ts (hunk 6 above)
Comment (aicioara) [inline src/plots/xyPlot.ts]: Renamed to `lazyDomainChange` and added documentation

File: src/plots/xyPlot.ts (hunk 7 above)
Comment (aicioara) [inline src/plots/xyPlot.ts]: Fixed. It failed because other (now) known bugs https://github.com/palantir/plottable/issues/2396

File: src/plots/xyPlot.ts (hunk 8 above)
Comment (aicioara) [inline src/plots/xyPlot.ts]: it's not a render, hence called it `lazyDomainChangeCallback`

**Comment by aicioara** [comment]
Used the constructor as a closure. Could not really think of another way to do it nicely. This just means the constructor code is bigger

**Comment by aicioara** [comment]
Writing an RFC for this. Code-wise, I think it is okay now.

File: src/plots/xyPlot.ts [hunk 9]
```diff
@@ -135,15 +237,28 @@ export class XYPlot<X, Y> extends Plot {
     var adjustCallback = key === XYPlot._X_KEY ? this._adjustYDomainOnChangeFromXCallback
                                                : this._adjustXDomainOnChangeFromYCallback;
     scale.onUpdate(adjustCallback);
+
+    if (this._lazyDomainChange) {
+      scale.offUpdate(this._renderCallback);
```
Comment (jtlan) [inline src/plots/xyPlot.ts]: Instead of doing this on/off switch, wouldn't it make more sense if `_renderCallback`'s internal behavior switched on `_lazyDomainChange`?

Comment (aicioara) [inline src/plots/xyPlot.ts]: The issue is that we need to not change the reference of the `this._renderCallback`. This means that we cannot plug and play a new `_rendererCallback`. This means that some sort of switch has to go inside the _rendererCallback. This means that the `Plot` needs to know about the fast version (since the `rendererCallback` is defined there)

Comment (aicioara) [inline src/plots/xyPlot.ts]: Alternative is to do this switch inside the `lazyDomainChange`, and fetch the scales from inside there

Comment (jtlan) [inline src/plots/xyPlot.ts]: Could we have a separate function that switches between them only for the X and Y `Scale`s?
```Typescript
this._xScaleRenderCallback = (scale) => {
  if (this.lazyDomainChange()) {
    this._lazyDomainChangeCallbackX(scale);
  } else {
    this._renderCallback(scale);
  }
};

...

protected _installScaleForKey(scale: Scale<any, any>, key: string) {
  if (key === XYPlot._X_KEY) {
    scale.onUpdate(this._xScaleRenderCallback);
  } else {
    super._installScaleForKey(scale, key);
  }
}
```

Not sure this is better, though.

Comment (aicioara) [inline src/plots/xyPlot.ts]: We'll miss important stuff from the parent class (like `scale.addIncludedValuesProvider(this._includedValuesProvider);`

Comment (jtlan) [inline src/plots/xyPlot.ts]: Alternatively:
```Typescript
this._renderCallback = (scale) => {
  if (scale === this.x().scale && this.lazyDomainChange()) {
    this._lazyDomainChangeCallbackX(scale);
  } else if (scale === this.y().scale && this.lazyDomainChange()) {
    this._lazyDomainChangeCallbackY(scale);
  } else {
    this._render();
  }
};
```

Comment (aicioara) [inline src/plots/xyPlot.ts]: That looks great!

File: src/plots/xyPlot.ts [hunk 10]
```diff
@@ -53,10 +152,6 @@ export class XYPlot<X, Y> extends Plot {
       this._updateYExtentsAndAutodomain();
     }
 
-    if (xScale != null) {
-      xScale.onUpdate(this._adjustYDomainOnChangeFromXCallback);
```
Comment (jtlan) [inline src/plots/xyPlot.ts]: ?? Why was this removed

Comment (aicioara) [inline src/plots/xyPlot.ts]: Redundant. `this.bindProperty()` -> `_bind()`-> `_installScaleForKey()` -> `scale.onUpdate(adjustDomainCallback)`

Comment (jtlan) [inline src/plots/xyPlot.ts]: Neat.

**Comment by jtlan** [comment]
Pending rename of API point for enabling this.

**Comment by aicioara** [comment]
Added the performance configuration concept here If we decide not to go for it, I can always reset a couple of commits back, once we figure out how to name things

**Comment by aicioara** [comment]
Blocking until we find better naming

**Comment by crmorford** [comment]
I'm seeing very strange behavior from horizontal bar plots, it's hard to figure out what's happening. On this test, when you pan, the size of the bars is changing unpredictably. http://jsfiddle.net/2ctcjx83/

On the quicktests (about to push) categoryAxis_timeAxis has a panning lag, which I didn't expect. The bars are 'jumping' to their end position when you stop dragging.

**Comment by aicioara** [comment]
This is the same issue as here https://github.com/palantir/plottable/pull/2344
Since this is the third time this feature is signaled as a flag (twice by @crmorford and once by Slate), I believe we should eliminate the feature.
The issue was that the test was containing a 0-length bar which made it harder to see what is going on. Amended the test here http://jsfiddle.net/2ctcjx83/4/ 

For completeness, the feature that can be observed in the fiddle is the following: `Plots.Bar` tries to fit all its bars in the given space if possible. As you approach upper or lower margins the behavior can be observed. This feature cannot exist in the fast version of panzoom. 

@crmorford Let me know if the test makes sense now

**Comment by aicioara** [comment]
The categoryAxis_timeAxis is an issue. Solving now. It probably has to do with grouping

**Comment by aicioara** [comment]
Fixed

**Comment by crmorford** [comment]
Oh yeah, the 0-length bar made it way harder to understand. Now it's much more predictable.

**Comment by crmorford** [comment]
And categoryAxis_timeAxis looks good now.

**Comment by aicioara** [comment]
I had a followup on this issue.

Issue: Scaling on a SVG influences both the size of the symbols and the distance between them. What we want is to influence the distance between symbols while leaving the symbol sizes untouched. 

@dcervelli suggested that we can maybe apply a CSS3 transform on the symbol class, which will do an inverse scaling on the symbol itself and that would solve the problem. There were two issues with that:
- CSS3 transforms on SVG elements are not supported in IE [known issue](https://connect.microsoft.com/IE/feedbackdetail/view/1046003/css-transforms-feature-detection-not-working-on-svg-elements)
- Applying CSS3 transforms in Chrome overwrites the SVG transform attributes (so we lose the position of the points)
